### ISSUE TITLE
Crafting List

### ISSUE THREAD

**Issue by DevDobariya** [refinedmods/refinedstorage2#369]
There should be a List item in which we can add crafting recipes and also amount of each item that is added. Like if I want to craft 48 SPS Casing(Mekanism mod) and 8 SPS Port(Mekanism Mod) then just put the item and name the list and number of it in the list and then add the list to a station that stores the lists and when the list is clicked it will showcase the to craft menu and when pressed start crafting it will craft all the contents in the list.

I know this might be a little complex to code but it would be helpful if someone wants to make some Multiblocks more than one time.

**Comment by lucasser** [comment]
If you do implement this, you can have an online library of the recipe lists for multiblocks for well known mods and the player can access the lists from the list grid.

**Comment by jonatjano** [comment]
Since RS2 autocrafting won't be available before minecraft 1.20 (most probably), maybe there could be a bundle pattern which like a [Bucket pattern #168](https://github.com/refinedmods/refinedstorage2/issues/168) will take a [vanilla 1.20 Bundle](https://minecraft.fandom.com/wiki/Bundle) and fill it with the requested items
The limit of 64 items seems good for most use cases

We can then hope that mods adding backpack make them work as bundle or for a big bundle mod (which I'm sure will exist) to allow bigger auto bundle

**Comment by raoulvdberge** [comment]
I would create a new type of pattern - the batch pattern. It can be encoded from the pattern grid as its own unique pattern type.

A batch pattern will have a name & a list of sub-ingredients with an amount.

A batch pattern is displayed from the grid as its own unique resource (next to existing items/fluids). If you click on it, you will request autocrafting for the entire list.

In the Autocrafting Monitor, you would only see 1 task, for the entire batch, and see "subtasks".

Open questions:

- Where in the Grid are the batch patterns shown?
- What happens if you request from the batch and want to get the crafted resources out? We could just dump all the outputs in storage, but I feel like players would like to easily retrieve the outputs at once.

**Comment by jonatjano** [comment]
> Where in the Grid are the batch patterns shown?

In addition to the name the pattern could take an item to serve as icon and use that item in the grid

> What happens if you request from the batch and want to get the crafted resources out? We could just dump all the outputs in storage, but I feel like players would like to easily retrieve the outputs at once.

If the crafter with the pattern point into an inventory, push into it, otherwise put it back in the storage, with maybe information on that in the batch pattern item extended description

**Comment by raoulvdberge** [comment]
> In addition to the name the pattern could take an item to serve as icon and use that item in the grid

Yes, but what about the ordering? Just show them all at the end?

> If the crafter with the pattern point into an inventory, push into it, otherwise put it back in the storage, with maybe information on that in the batch pattern item extended description

Just scattering resources part of the batch in storage is what I'm trying to avoid.

Putting them in an inventory doesn't make sense - autocrafters only push inputs to external inventories and a batch pattern can never be a valid input for another pattern. It's really a top level pattern.

**Comment by jonatjano** [comment]
> Yes, but what about the ordering? Just show them all at the end?

I don't remember all the ordering options but if the order is by name, use the name, if the order is by count, always consider that there are 0 of them (or 1 if it makes the logic easier)


> Just scattering resources part of the batch in storage is what I'm trying to avoid.

> Putting them in an inventory doesn't make sense - autocrafters only push inputs to external inventories and a batch pattern can never be a valid input for another pattern. It's really a top level pattern.

It does make sense, the player can put the auto crafter on a chest, there is even one machine which takes batches of item and process them : create's schematicanon
Batch pattern cannot be input to another pattern, but it doesn't mean they are not input to further processing

**Comment by raoulvdberge** [comment]
> It does make sense, the player can put the auto crafter on a chest, there is even one machine which takes batches of item and process them : create's schematicanon
Batch pattern cannot be input to another pattern, but it doesn't mean they are not input to further processing

From a player perspective this is confusing; the autocrafter never inserts "final outputs" back into its connected inventory - until now?

**Comment by jonatjano** [comment]
I know some players emulate batch pattern by doing processing pattern resulting in a renamed piece of paper, put the auto crafter against a chest and when all the items are put in the chest you either need to insert the correct named paper into the storage via an importer or cancel the task manually
There was a discussion about this once on discord and we were like 5 person using a workaround similar to that

**Comment by raoulvdberge** [comment]
Yep, but if we implement this feature then the processing pattern workaround is not needed. That's the perspective where I'm coming from.

**Comment by lucasser** [comment]
I Also like pointing a crafter at an inventory with a renamed paper, I think if the batch pattern worked like a processing pattern, except it would come up as an "batch pattern" in grid, and would complete as soon as the items get pushed into the inventory that would be sufficient.

**Comment by raoulvdberge** [comment]
The idea is to *not* make this behave like a processing pattern, since doing the "renamed paper" workaround was a workaround to begin with, and I like to design the feature without thinking about workarounds.

**Comment by uhloin** [comment]
> * Where in the Grid are the batch patterns shown?

Looks like there is no sense to put it to some autocrafter, if pattern really contains items made by different autocrafters.
But if do not put it to auto-crafter, then where? Separate type? They we need special container only for such rare type.
As player I would prefer see it as normal pattern in the chest (may be with a bit different icon) with normal behavior (ctrl+click to start auto-craft). Why normal behavior? Because if some item in the chest launches something big when I just want to get all items from chest it would be annoying.

> * What happens if you request from the batch and want to get the crafted resources out? We could just dump all the outputs in storage, but I feel like players would like to easily retrieve the outputs at once.

I think it should dump, but make easy to find results in storage. May be it's similar to feature to find recently crafted items. If implement it, then this closes this question too.

**Comment by jonatjano** [comment]
You could also make a batch pattern driver which takes one(or many) batch pattern and act as a chest where player can only remove items, that block could have different tabs like the auto crafting monitor with one tab per batch of items

**Comment by lucasser** [comment]
Alternatively, the batch pattern oten works like a bundle/create package/shulker, where if you try to grab it from grid and you have all the resources, they get bundled up into one item which you can then open/decompress to get your things
